File: Java-Kotlin/src/main/java/tree/_102.java
package tree;

import common.datastructure.TreeNode;

import java.util.*;

public class _102 {

    //BFS solution
    public List<List<Integer>> levelOrder(TreeNode root) {
        if (root == null) {
            return Collections.emptyList();
        }

        List<List<Integer>> result = new ArrayList<>();

        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);

        while (!queue.isEmpty()) {
            int levelSize = queue.size();
            List<Integer> currLevel = new ArrayList<>();
            for (int i = 0; i < levelSize; i++) {
                TreeNode node = queue.poll();
                currLevel.add(node.val);
                if (node.left != null) {
                    queue.offer(node.left);
                }
                if (node.right != null) {
                    queue.offer(node.right);
                }
            }
            result.add(currLevel);
        }

        return result;
    }

    // DFS
    public List<List<Integer>> levelOrder1(TreeNode root) {
        if (root == null) {
            return Collections.emptyList();
        }

        List<List<Integer>> result = new ArrayList<>();

        helper(root, result, 0);

        return result;
    }

    private void helper(TreeNode node, List<List<Integer>> result, int level) {
        if (result.size() == level) {
            result.add(new ArrayList<>());
        }

        result.get(level).add(node.val);

        if (node.left != null) {
            helper(node.left, result, level + 1);
        }

        if (node.right != null) {
            helper(node.right, result, level + 1);
        }
    }
}

File: Go-Solutions/102.go
package main
// Problem link : https://leetcode-cn.com/problems/binary-tree-level-order-traversal/
// DFS
func levelOrder(root *TreeNode) [][]int {
	if root == nil {
		return [][]int{}
	}
	result := make([][]int, 0)
	helperLevelOrder(root, &result, 0)
	return result
}

func helperLevelOrder(root *TreeNode, result *[][]int, level int) {
	if len(*result) == level {
		*result = append(*result, []int{})
	}

	(*result)[level] = append((*result)[level], root.Val)

	if root.Left != nil {
		helperLevelOrder(root.Left, result, level + 1)
	}

	if root.Right != nil {
		helperLevelOrder(root.Right, result, level + 1)
	}
}

// bfs
func levelOrder1(root *TreeNode) [][]int {
	if root == nil {
		return [][]int{}
	}
	result := make([][]int, 0)

	queue := make([]*TreeNode, 0)
	queue = append(queue, root)

	for len(queue) != 0 {
		size := len(queue)
		currentLevel := make([]int, 0)
		for i := 0; i < size; i++ {
			// 相当于java里面的poll操作
			node := queue[0]
			queue = queue[1:]
			if node.Left != nil {
				queue = append(queue, node.Left)
			}
			if node.Right != nil {
				queue = append(queue, node.Right)
			}
			currentLevel = append(currentLevel, node.Val)
		}
		result = append(result, currentLevel)
	}
	return result
}
